### DISCUSSION TITLE
Stream a file chunk by chunk

### DISCUSSION THREAD

**Discussion by vincentLock21** [maennchen/ZipStream-PHP#178] [Q&A] [answered]
Hello,

I am currently using `ZipStream` in my project, but I encountered a fairly specific case and I wanted to know if there was a solution available to me.

I use the `addFileFromStream` function because I indeed need to stream, but the problem is that I am not recovering whole files but chunks from S3. So I would have liked to know **If it was possible to stream a file chunk by chunk using ZipStream** ?

For the **options** configuration :

```php
$options = new Archive();
$options->setContentType('application/zip');
$options->setContentDisposition('attachment; filename="test.zip"');
$options->setZeroHeader(true);
$options->setComment('test zip file.');
$options->setSendHttpHeaders(true);
```

For now, I'm going through temporary **in-memory storage** :
```php
$stream = fopen('php://memory', 'ab+');
```

The chunk being encrypted, I decrypt it and transform the returned character string into a resource to be able to pass it to the function.

```php
$streamRead = \openssl_decrypt(
  stream_get_contents(fopen($s3path, 'rb')),
  /* ... */
);
fwrite($stream, $streamRead);
rewind($stream);
$zip->addFileFromStream($fileName, $stream);
```

Unfortunately, by doing this each chunk overwrites the previous one instead of ending at the end of the file.

Thanks for your help !

**Comment by maennchen** [comment, marked as answer]
It should probably be possible to implement a custom class implementing PSR7 StreamInterface to do this.

**EDIT:**

This could work something like that (completely untested pseudo code):

```php
use Psr7\AppendStream;
use Psr7\Utils;

$zip = new ZipStream('test.zip', $options);

foreach ($s3keys as $key => $value) {
  $fileName = $value;

  // $chunks = get chunks from S3;

  $fileStream = new AppendStream(array_map(function($chunk) using($bucket, $key, $value) {
    $path = sprintf('s3://%s/%s/%s/%s', $bucket, $key, $chunk->getChunk(), basename($value));

    return Utils::streamFor(openssl_decrypt(fopen($path, 'rb'), /* ... */));
  }, $chunks));

  $zip->addFileFromPsr7Stream($fileName, $fileStream);
}

$zip->finish();
```

**Reply by vincentLock21** [reply]
Thanks for your response ! Do you have an example for this ?

**Reply by maennchen** [reply]
Here you can see the implementation of a normal Stream: https://github.com/guzzle/psr7/blob/master/src/Stream.php

You could probably read the chunk in the read function and then start reading from the next as soon as you reach eol.

**Reply by vincentLock21** [reply]
Thanks again for your quick answer ! :)
I tried a solution with the following code (StreamService is my custom class who implements StreamInterface, I have copy/paste your example class Stream) :

```php
      return new StreamedResponse(function() use($s3keys) {
            $options = new Archive();
            $options->setContentType('application/zip');
            $options->setContentDisposition('attachment; filename="test.zip"');
            $options->setZeroHeader(true);
            $options->setComment('test zip file.');
            $options->setSendHttpHeaders(true);

            $zip = new ZipStream('test.zip', $options);

            foreach ($s3keys as $key => $value) {
                $fileName = $value;

                // chunks = get chunks from S3;

                $stream = fopen('php://memory', 'ab+');
                $s = new StreamService($stream);

                foreach ($chunks as $chunk) {
                    $s3path = "s3://" . $_ENV['bucket'] . "/" . $key . '/' . $chunk->getChunk() . '/' . \basename($value);
                    $streamRead = \openssl_decrypt(fopen($s3path, 'rb'), XXX);
                    $s->write($streamRead);
                    $zip->addFileFromPsr7Stream($fileName, $s);
                }
            }

            $zip->finish();
        });
```

Unfortunately, it's not working and I get an archive with the rights files and paths, but files are empty (0 bytes). My variable stream is probably not well used, but I don't understand what I'm doing wrong. Do you have an idea ?

Thanks a lot for your help !

**Reply by dudz94** [reply]
Hello guys,

First, thanks for this bundle... It working very great where the files are stored in one file ! 

I have exactly the same issue. I want to stream multiple files but my files stored in S3 are in chunks.
I try many things, try to implement a normal Stream, tried the @vincentLock21's code but nothing works.

@maennchen do you have an idea ?

**Reply by maennchen** [reply]
The code example @vincentLock21 posted unfortunately has nothing to do with what I suggested.

One needs to take the `StreamInterface` and write one that can handle multiple streams. 

It seems like someone has done something similar before:

- https://github.com/guzzle/psr7/blob/master/src/AppendStream.php
- https://github.com/guzzle/psr7#appendstream

This could work something like that (completely untested pseudo code):

```php
use Psr7\AppendStream;
use Psr7\Utils;

$zip = new ZipStream('test.zip', $options);

foreach ($s3keys as $key => $value) {
  $fileName = $value;

  // $chunks = get chunks from S3;

  $fileStream = new AppendStream(array_map(function($chunk) using($bucket, $key, $value) {
    $path = sprintf('s3://%s/%s/%s/%s', $bucket, $key, $chunk->getChunk(), basename($value));

    return Utils::streamFor(openssl_decrypt(fopen($path, 'rb'), /* ... */));
  }, $chunks));

  $zip->addFileFromPsr7Stream($fileName, $fileStream);
}

$zip->finish();
```

**Reply by dudz94** [reply]
Hello @maennchen !

Thanks a lot for your help. Code you provide works perfectly.
The only things is the file is write before on disk before the download start. I have a lot of big files (100Go+) and the filesystem cannot write it before send to the browser.

Do you know if there is a way to send the file chunk by chunk without get the all chunks in the server ? 

Thanks again for your help !

**Reply by maennchen** [reply]
Could you try with `$options->setFlushOutput(true);`

**Reply by vincentLock21** [reply]
Hello @maennchen ! Thanks for all !

I've tried with `$options->setFlushOutput(true);` but nothing. I think this is because the ZipStream outpout is `fopen('php://output', 'wb');` by default , and `Utils::streamFor` write on `fopen('php://temp', 'r+');`.

I tried `$options->setOutputStream(fopen('php://temp', 'r+'))` and :
`$stream = $zip->opt->getOutputStream();`
`fwrite($stream, $resource);`
`fseek($stream, 0);`
But the result is an empty archive.

I also tested to use, to overcome this, the `send()` function of ZipStream, which writes to the ZipStream's default outpout, and there I have the chunks which are streamed one after the other. When I open my Mac Finder I can see the zip with the correct size. But impossible to open it. 

Here is my code :

            $options = new Archive();
            $options->setFlushOutput(true);
            $options->setSendHttpHeaders(true);

            $zip = new ZipStream('test.zip', $options);

            foreach ($s3keys as $key => $value) {
                $fileName = $value;
                $chunks = //get chunks s3;

                $fileStream = new AppendStream(array_map(function($chunk) use ($key, $value, $zip) {
                    $s3path = XXX;

                    $str = \openssl_decrypt(
                        stream_get_contents(fopen($s3path, 'rb')),
                        XXX,
                        ....
                    );
                    $zip->send($str);

                    return $zip->opt->getOutputStream();

                }, $chunks));

                $zip->addFileFromPsr7Stream($fileName, $fileStream);
            }

            $zip->finish();`

Did I understand the use of the `send()` function correctly? Chunks are well streamed one by one but the archive may not be correctly reconstituted?

Thank you again for your time!

**Reply by maennchen** [reply]
@vincentLock21 You do not need `ZipStream::send`. This is to send data to the output stream and I'm not even sure right now why it even is public...

ZipStream will start writing into the output stream (default `php://output`) as soon as you start calling `addFile*` functions.

ZipStream doe not save anything to disk by itself. That only happens if the output stream is saved to the disk (for example if the output is not flushed or if specifying another target stream) or if the input streams are on disk.

In this case, I would guess that this is something the S3 SDK does.

My time is quite limited at the moment. So to dig around the S3 SDK I do not have the resources to help, sorry.

**Reply by vincentLock21** [reply]
Ok I'll watch this ! Thanks for all ! 😃
